### DISCUSSION TITLE
Explore the possibility of using Hpy API instead of C API

### DISCUSSION THREAD

**Discussion by saidctb** [pyccel/pyccel#1043] [Ideas]
The goal of [Hpy](https://github.com/hpyproject/hpy) is to design a better API for extending Python in C.
 You can find here https://github.com/hpyproject/hpy/blob/master/docs/api.rst a simple example to do it.
The goal of hpy is to support other Python implementations (e.g. PyPy, GraalPython, Jython, IronPython, etc.), this will help us extend Pyccel to other Python implementations for free.

**Comment by EmilyBourne** [comment]
Note the following in the [docs](https://hpy.readthedocs.io/en/latest/overview.html#current-status-and-roadmap)

> At the moment of writing, HPy is still in its early stages of development. The following milestones have been reached:
> 
> 1. it is possible to write extensions which expose module-level functions, with all the various kinds of calling conventions
> 2. there is a limited support for argument parsing (only a couple of basic types actually work)
> 3. there is support for raising and catching exceptions
> 4. it is possible to choose between the CPython ABI and the HPy Universal ABI when compiling an extension module
> 5. extensions compiled with the CPython ABI work out of the box on CPython
> 6. it is possible to load HPy Universal extensions on CPython, thanks to the hpy.universal package
> 7. it is possible to load HPy Universal extensions on PyPy (using the PyPy hpy branch)
> 8. it is possible to load HPy Universal extensions on GraalPython
> 
> However, there is still a long road before HPy is usable for the general public. In particular, the following features are on our roadmap but have not been implemented yet:
> 
> 9. it is not possible to write custom types (like NumPy’s ndarray) in C. There is already a WIP branch to address this issue
> 10. only a handful of the original Python/C functions have been ported to HPy. Porting most of them is straighforward, so for now the priority is to work on the “hard” features to prove that the HPy approach works, and we will port new functions as needed
> 11. the debug mode simply does not exist (yet!)
> 12. there is no standard/easy way to integrate HPy into a distutils/setuptools based workflow. The only way to compile HPy extensions right now is to manually clone the git repo and tweak your setup.py. Eventually, we plan to offer a workflow which integrates seamlessly with pip, setuptools, etc.
> 13. there is no integration with Cython. The medium-term plan is to extend Cython to automatically generate HPy-compatible C code

For us the problems currently are mostly 2. and 9. but also possibly 12

**Comment by saidctb** [comment]
yes, it's still in its early stages, I wonder if we can do  a hybrid approach, where we use the C API  where Hpy doesn't work, like in 2, and for the 9 we can wait for them to fix it, and 12 can possibly be fixed,
 From the example that they have,  we probably only need a dictionary that maps the C API to Hpy API  functions.
